### PR TITLE
Introduce "metric iterator"

### PR DESCRIPTION
**Pull request by alonre24**
Add a new generic iterator that only a list of element with two fields: doc id, and a metric (number).
It's corresponding `RSIndexResult` type is what we used before as `RSDistanceRecord` (renamed to `RSMetricRecord`).

Implement vector range query iterator as a metric iterator. Tests are WIP.

### CODE REVIEW

File: src/metric_iterator.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  MetricIterator *mr = ctx;
+
+  // Set the item that we read in the current RSIndexResult
+  *hit = mr->base.current;
+  (*hit)->docId = mr->lastDocId = mr->idsList[mr->curIndex];
+  (*hit)->metric.value = mr->metricList[mr->curIndex];
+  (*hit)->metric.metricField = mr->fieldName;
+
+  mr->curIndex++;
+  if (mr->curIndex == mr->resultsNum) {
```
Comment (ashtul) [inline src/metric_iterator.c:44]: add comments about advancing `curIndex` and reaching EOF. 

`IITER_SET_EOF` `IITER_CLEAR_EOF`

File: src/metric_iterator.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+static int MR_SkipTo(void *ctx, t_docId docId, RSIndexResult **hit) {
+  if (!MR_HasNext(ctx)) {
+    return INDEXREAD_EOF;
+  }
+  MetricIterator *mr = ctx;
+  t_docId cur_id = mr->idsList[mr->curIndex];
+  while(mr->curIndex < mr->resultsNum) {
+    if (docId > cur_id) {
+      // consider binary search for next value
```
Comment (ashtul) [inline src/metric_iterator.c:58]: consider skipping 1,2,4,8,,, on the search.
slightly better then binary search

File: src/metric_iterator.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include "index_iterator.h"
+
+typedef enum {
+  VECTOR_DISTANCE,
+} Metric;
+
+typedef struct {
```
Comment (ashtul) [inline src/metric_iterator.h:9]: Let's reorder and comment

File: src/query_parser/v2/parser.y [hunk 4]
```diff
@@ -842,11 +842,7 @@ geo_filter(A) ::= LSQB param_num(B) param_num(C) param_num(D) param_term(E) RSQB
 
 query ::= expr(A) ARROW LSQB vector_query(B) RSQB . { // main parse, hybrid query as entire query case.
   setup_trace(ctx);
-  switch (B->vn.vq->type) {
-    case VECSIM_QT_KNN:
-      B->vn.vq->knn.order = BY_SCORE;
-      break;
-  }
+  RedisModule_Assert(B->vn.vq->type == VECSIM_QT_KNN);
```
Comment (ashtul) [inline src/query_parser/v2/parser.y:845]:
```suggestion
  RS_LOG_ASSERT(B->vn.vq->type == VECSIM_QT_KNN, "vector_query must be KNN");
```

Please use `RS_LOG_ASSERT`.

**Review by ashtul — APPROVED** [review]
Looks good.
Minor styling comments

File: src/metric_iterator.c (hunk 2 above)
Comment (alonre24) [reply src/metric_iterator.c:58]: I think this is also called a binary search. I updated the comment anyway

File: src/metric_iterator.h (hunk 3 above)
Comment (alonre24) [reply src/metric_iterator.h:9]: what should be reordered?

File: tests/pytests/test_vecsim.py [hunk 5]
```diff
@@ -1515,3 +1599,79 @@ def test_timeout_reached():
                 env.assertEqual(res[0], timeout_expected)
 
             conn.flushall()
+
+
+def test_range_query_basic():
+    env = Env(moduleArgs='DEFAULT_DIALECT 2 ON_TIMEOUT FAIL')
```
Comment (GuyAv46) [inline tests/pytests/test_vecsim.py:1605]: Why `ON_TIMEOUT FAIL`?

File: tests/pytests/test_vecsim.py [hunk 6]
```diff
@@ -1515,3 +1599,79 @@ def test_timeout_reached():
...
+                                       'RETURN', 0, 'LIMIT', 0, n)
+            env.assertEqual(res[0], 499)
+            for i, doc_id in enumerate(res[1:]):
+                env.assertEqual(str(500 + i + 1), doc_id)  # results should be sorted by id (by default)
+
+            conn.flushall()
+
+
+def test_range_query_basic_random_vectors():
+    env = Env(moduleArgs='DEFAULT_DIALECT 2 ON_TIMEOUT FAIL')
```
Comment (GuyAv46) [inline tests/pytests/test_vecsim.py:1648]: Why `ON_TIMEOUT FAIL`?

**Review by GuyAv46 — COMMENTED** [review]
Looks good
I'm missing range queries for multi-value

**Pull request merged**